crates/typst-library/src/text/shift.rs

```diff
@@ -108,107 +120,55 @@ pub struct SuperElem {
-        let body = self.body.clone();
-
-        if self.typographic(styles) {
-            if let Some(text) = convert_script(&body, false) {
-                if is_shapable(engine, &text, styles) {
-                    return Ok(TextElem::packed(text));
-                }
-            }
-        };
-
-        Ok(body
-            .styled(TextElem::set_baseline(self.baseline(styles)))
-            .styled(TextElem::set_size(self.size(styles))))
+    fn show(&self, _: &mut Engine, styles: StyleChain) -> SourceResult<Content> {
+        let outer_text_size = TextElem::size_in(styles);
+        Ok(self
+            .body
+            .clone()
+            .styled(TextElem::set_subperscript(Some(ScriptSettings {
+                typographic: self.typographic(styles),
+                shift: self.baseline(styles).map(|l| -l.to_em(outer_text_size)),
+                size: self.size(styles).map(|t| t.0.to_em(outer_text_size)),
+                kind: ScriptKind::Super,
+            }))))
     }
 }
 
-/// Find and transform the text contained in `content` to the given script kind
-/// if and only if it only consists of `Text`, `Space`, and `Empty` leaves.
-fn convert_script(content: &Content, sub: bool) -> Option<EcoString> {
-    if content.is::<SpaceElem>() {
-        Some(' '.into())
-    } else if let Some(elem) = content.to_packed::<TextElem>() {
-        if sub {
-            elem.text.chars().map(to_subscript_codepoint).collect()
-        } else {
-            elem.text.chars().map(to_superscript_codepoint).collect()
-        }
-    } else if let Some(sequence) = content.to_packed::<SequenceElem>() {
-        sequence
-            .children
-            .iter()
-            .map(|item| convert_script(item, sub))
-            .collect()
-    } else {
-        None
-    }
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ScriptKind {
+    Sub,
+    Super,
 }
 
-/// Checks whether the first retrievable family contains all code points of the
-/// given string.
-fn is_shapable(engine: &Engine, text: &str, styles: StyleChain) -> bool {
-    let world = engine.world;
-    for family in TextElem::font_in(styles) {
-        if let Some(font) = world
-            .book()
-            .select(family.as_str(), variant(styles))
-            .and_then(|id| world.font(id))
-        {
-            let covers = family.covers();
-            return text.chars().all(|c| {
-                covers.is_none_or(|cov| cov.is_match(c.encode_utf8(&mut [0; 4])))
-                    && font.ttf().glyph_index(c).is_some()
-            });
+impl ScriptKind {
+    pub const fn default_metrics(self) -> ScriptMetrics {
+        match self {
+            Self::Sub => ScriptMetrics::default_subscript(),
+            Self::Super => ScriptMetrics::default_superscript(),
         }
     }
 
-    false
-}
+    pub const fn read_metrics(self, font_metrics: &FontMetrics) -> ScriptMetrics {
```

Comment:
we could return `&ScriptMetrics` here and for the defaults have them in `static`s and return references to those.